ipaserver/plugins/servicedelegation.py

```diff
@@ -314,10 +342,21 @@ def post_callback(self, ldap, completed, failed, dn, entry_attrs,
             for name in names:
                 if not name:
                     continue
-                name = normalize_principal(name)
                 try:
-                    if name in entry_attrs.get(self.principal_attr, []):
-                        entry_attrs[self.principal_attr].remove(name)
+                    n = kerberos.Principal(name, realm=self.api.env.realm)
+                except ValueError:
+                    raise errors.ValidationError(
+                        name='principal',
+                        reason=_("Malformed principal: %(error)s") % dict(
+                            error=str(e)))
+
+                if len(n.components) == 1 and (n.components[0][-1] != '$'):
+                    nprincipal = 'host/' + unicode(n)
+                else:
+                    nprincipal = unicode(n)
+                try:
+                    if nprincipal in entry_attrs.get(self.principal_attr, []):
+                        entry_attrs[self.principal_attr].remove(nprincipal)
```

Comment:
This block looks very similar to the modified block below. How about moving the code into a shared helper function?